### PR TITLE
[SPARK-25408] Move to mode ideomatic Java8

### PR DESCRIPTION
**Pull request by Fokko**
While working on another PR, I noticed that there is quite some legacy Java in there that can be beautified. For example the use og features from Java8, such as:
- Collection libraries
- Try-with-resource blocks

No code has been changed

What are your thoughts on this?

This makes code easier to read, and using try-with-resource makes is less likely to forget to close something.

## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: common/kvstore/src/test/java/org/apache/spark/util/kvstore/DBIteratorSuite.java
Comment (TomaszGaweda) [inline common/kvstore/src/test/java/org/apache/spark/util/kvstore/DBIteratorSuite.java]: If you are already changing this, I would  also suggest to add static imports to shorten the code

**Comment by TomaszGaweda** [comment]
I like this idea :) +1 as it's only refactor, without logic change

File: common/kvstore/src/test/java/org/apache/spark/util/kvstore/DBIteratorSuite.java
Comment (srowen) [reply common/kvstore/src/test/java/org/apache/spark/util/kvstore/DBIteratorSuite.java]: I don't think this sort of thing is worth changing. I know, IntelliJ suggests it. Unless the mutability is an issue, I'd leave it as a shorter idiom. I wouldn't use static imports here personally.

there's also a very small cost of changes in that they create potential merge conflicts for other changes later, so I tend to have a very low but finite minimum bar for value of code scrubbing like this.

File: common/network-common/src/test/java/org/apache/spark/network/ChunkFetchIntegrationSuite.java
Comment (srowen) [inline common/network-common/src/test/java/org/apache/spark/network/ChunkFetchIntegrationSuite.java]: Nit: space after try

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolverSuite.java
Comment (srowen) [inline common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolverSuite.java]: Same comment about space above; I'd also indent the continuation 4 spaces for clarity

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleIntegrationSuite.java
Comment (srowen) [inline common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleIntegrationSuite.java]: Indent is now too deep here. I have the same general kind of doubt here.. it's touching a lot of lines for little actual gain. Still, I'd like to be able to improve code a bit here and there. If this is only going to master and Spark 3, the back-port issue lessens, because it's more unlikely to backport from 3.x to 2.x.

File: core/src/test/java/org/apache/spark/launcher/SparkLauncherSuite.java
Comment (srowen) [inline core/src/test/java/org/apache/spark/launcher/SparkLauncherSuite.java]: Don't collapse these

File: launcher/src/main/java/org/apache/spark/launcher/AbstractAppHandle.java
Comment (srowen) [inline launcher/src/main/java/org/apache/spark/launcher/AbstractAppHandle.java]: I wouldn't bother with this; we don't really do it consistently, it's not a JDK standard class, and it doesn't really save much, while adding to the dependency on commons/io

**Review by srowen — CHANGES_REQUESTED** [review]

File: launcher/src/main/java/org/apache/spark/launcher/AbstractAppHandle.java
Comment (vanzin) [reply launcher/src/main/java/org/apache/spark/launcher/AbstractAppHandle.java]: This library should not have any non-JRE dependencies.

File: common/kvstore/src/test/java/org/apache/spark/util/kvstore/DBIteratorSuite.java
Comment (Fokko) [reply common/kvstore/src/test/java/org/apache/spark/util/kvstore/DBIteratorSuite.java]: I agree, I don't really care about moving to `Collections`.

File: launcher/src/main/java/org/apache/spark/launcher/AbstractAppHandle.java
Comment (Fokko) [reply launcher/src/main/java/org/apache/spark/launcher/AbstractAppHandle.java]: I wan't aware of this requirement, reverted the change.

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleIntegrationSuite.java
Comment (HyukjinKwon) [inline common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleIntegrationSuite.java]: Shell we stick to 2spaced indentation?

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleSecuritySuite.java
Comment (HyukjinKwon) [inline common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleSecuritySuite.java]: ditto for indentation

File: core/src/test/java/org/apache/spark/JavaJdbcRDDSuite.java
Comment (HyukjinKwon) [inline core/src/test/java/org/apache/spark/JavaJdbcRDDSuite.java]: indentation ..

File: common/network-shuffle/src/test/java/org/apache/spark/network/sasl/SaslIntegrationSuite.java
Comment (HyukjinKwon) [inline common/network-shuffle/src/test/java/org/apache/spark/network/sasl/SaslIntegrationSuite.java]: Looks unused.

**Review by HyukjinKwon — COMMENTED** [review]
was just looking through this out of curiosity. just few nits.

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolverSuite.java
Comment (kiszk) [inline common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolverSuite.java]: Would it be possible to add a space after `try`?

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolverSuite.java
Comment (kiszk) [inline common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolverSuite.java]: ditto

File: core/src/main/java/org/apache/spark/shuffle/sort/ShuffleExternalSorter.java
Comment (kiszk) [inline core/src/main/java/org/apache/spark/shuffle/sort/ShuffleExternalSorter.java]: nit: fix indentation

File: core/src/test/java/org/apache/spark/JavaJdbcRDDSuite.java
Comment (Fokko) [reply core/src/test/java/org/apache/spark/JavaJdbcRDDSuite.java]: This one is horrible yes

File: core/src/test/java/test/org/apache/spark/JavaAPISuite.java
Comment (kiszk) [inline core/src/test/java/test/org/apache/spark/JavaAPISuite.java]: nit: space after `try`

File: core/src/test/java/test/org/apache/spark/JavaAPISuite.java
Comment (kiszk) [inline core/src/test/java/test/org/apache/spark/JavaAPISuite.java]: ditto

**Comment by Fokko** [comment]
Thanks for the feedback guys, I was relying too much on Scalastyle :)

File: sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/CLIService.java
Comment (kiszk) [inline sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/CLIService.java]: Do we need `try` at line 149?

File: sql/catalyst/src/test/java/org/apache/spark/sql/catalyst/expressions/RowBasedKeyValueBatchSuite.java
Comment (kiszk) [inline sql/catalyst/src/test/java/org/apache/spark/sql/catalyst/expressions/RowBasedKeyValueBatchSuite.java]: nit: it would be good to delete this brank line.

File: sql/catalyst/src/test/java/org/apache/spark/sql/catalyst/expressions/RowBasedKeyValueBatchSuite.java
Comment (kiszk) [inline sql/catalyst/src/test/java/org/apache/spark/sql/catalyst/expressions/RowBasedKeyValueBatchSuite.java]: Can we put `key` and `value` inside `try`?

File: sql/catalyst/src/test/java/org/apache/spark/sql/catalyst/expressions/RowBasedKeyValueBatchSuite.java
Comment (kiszk) [inline sql/catalyst/src/test/java/org/apache/spark/sql/catalyst/expressions/RowBasedKeyValueBatchSuite.java]: ditto

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleIntegrationSuite.java
Comment (srowen) [reply common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleIntegrationSuite.java]: The scenario is something like: later, a line or two in this block is changed. That change is cherry-picked to a branch before this change here. It's a merge conflict, and choosing "theirs" overwrites your change here. It would have to be manually merged. If you know that this is all that has happened, sure, not hard. But it relies on the committer figuring that out and not missing another subtle change during the manual merge. That's definitely the most time-consuming part for me. For Spark 3.x vs 2.x I am less worried, and would like a freer hand to make small improvements. Within a major release I am reluctant.

File: sql/catalyst/src/test/java/org/apache/spark/sql/catalyst/expressions/RowBasedKeyValueBatchSuite.java
Comment (Fokko) [reply sql/catalyst/src/test/java/org/apache/spark/sql/catalyst/expressions/RowBasedKeyValueBatchSuite.java]: Good one

**Comment by Fokko** [comment]
Addressed comments and rebased onto master, no merge conflicts.

File: common/network-common/src/test/java/org/apache/spark/network/ChunkFetchIntegrationSuite.java
Comment (kiszk) [inline common/network-common/src/test/java/org/apache/spark/network/ChunkFetchIntegrationSuite.java]: nit: Can we put this line into `try` since this line was originally executed after `clientFactory.createClient`.

File: sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/CLIService.java
Comment (kiszk) [reply sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/CLIService.java]: Do we need `try` at line 149?

File: common/network-common/src/test/java/org/apache/spark/network/ChunkFetchIntegrationSuite.java
Comment (Fokko) [reply common/network-common/src/test/java/org/apache/spark/network/ChunkFetchIntegrationSuite.java]: No we can't, since the `res` is returned by the function :-)

Comment (Fokko) [reply common/network-common/src/test/java/org/apache/spark/network/ChunkFetchIntegrationSuite.java]: Any time

**Comment by Fokko** [comment]
@srowen Any incentive to move this forward? Or are PR's like these not appreciated? Let me know.

Most of the changes are cosmetic, but having <Java8 in the codebase give a bad example to newcomers to the Spark project. Also adding the `Closeable` to the `RowBasedKeyValueBatchSuite.java`:
https://github.com/apache/spark/pull/22399/files#diff-6c2c45f79666e2e52eb9f9411fa8b4baR49
makes the codebase a bit nicer in my opinion, since the class already has a `.close()` method, it makes sense to also implement the `Closeable` interface.

**Comment by srowen** [comment]
I think this is fine (though see https://github.com/apache/spark/pull/22399/files#r218121426 ). As a practical matter I might wait until we're done with the 2.4 release to start a lot of changes like this. It isn't tied to Spark 2.5 or 3.0 but might just be simpler to hold this for a bit, while we might be wanting to quickly backport a few changes from master to 2.4.

**Comment by Fokko** [comment]
Cool, makes sense. Thanks for the clarification @srowen

File: sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/CLIService.java
Comment (Fokko) [reply sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/CLIService.java]: Good one, thanks

**Comment by Fokko** [comment]
Rebased onto master

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Comment by gatorsmile** [comment]
It sounds like the build is broken after merging this PR

**Comment by gatorsmile** [comment]
https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Compile/job/spark-master-compile-maven-hadoop-2.6/8432/

**Comment by gatorsmile** [comment]
https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Compile/job/spark-master-compile-maven-hadoop-2.7/8521/

**Comment by cloud-fan** [comment]
Jenkins was not triggered in this PR, I'm reverting it. Please re-open it and make sure jenkins pass, thanks!

**Comment by srowen** [comment]
Oh no, I got mixed up and didn't realize this one hadn't actually had a jenkins run. Sorry about this, that is my mistake. I will follow up.

**Comment by srowen** [comment]
Sorry @Fokko can you recreate this change in a new PR, after addressing...
```
[error] /home/jenkins/workspace/spark-master-compile-maven-hadoop-2.6/sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/CLIService.java:149: error: incompatible types: try-with-resources not applicable to variable type
[error]     try (IMetaStoreClient metastoreClient = new HiveMetaStoreClient(hiveConf)) {
[error]                           ^
[error]     (IMetaStoreClient cannot be converted to AutoCloseable)
[error] /home/jenkins/workspace/spark-master-compile-maven-hadoop-2.6/sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/operation/OperationManager.java:200: error: incompatible types: try-with-resources not applicable to variable type
[error]     try (Operation operation = removeOperation(opHandle)) {
[error]                    ^
[error]     (Operation cannot be converted to AutoCloseable)
```

**Comment by Fokko** [comment]
Thanks @srowen for pointing out the errors. Weird that it did not come up as a merge conflict. Let me open a new PR.

**Comment by Fokko** [comment]
The tests passed earlier, how would it be possible that it would fail on master?

**Comment by HyukjinKwon** [comment]
Let me trigger it in the next PR.

**Comment by Fokko** [comment]
@HyukjinKwon I've opened a new PR under https://github.com/apache/spark/pull/22637. Would be nice if you can trigger Travis 👍

**Pull request closed without merging**
